ui/pb/progressbar_test.go

```diff
@@ -20,4 +20,114 @@
 
 package pb
 
-//TODO
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestProgressBarRender(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		options  []ProgressBarOption
+		expected string
+	}{
+		{[]ProgressBarOption{WithLeft(func() string { return "left" })},
+			"left [--------------------------------------]"},
+		{[]ProgressBarOption{WithConstLeft("constLeft")},
+			"constLeft [--------------------------------------]"},
+		{[]ProgressBarOption{
+			WithLeft(func() string { return "left" }),
+			WithProgress(func() (float64, string) { return 0, "right" }),
+		},
+			"left [--------------------------------------] right"},
+		{[]ProgressBarOption{
+			WithLeft(func() string { return "left" }),
+			WithProgress(func() (float64, string) { return 0.5, "right" }),
+		},
+			"left [==================>-------------------] right"},
+		{[]ProgressBarOption{
+			WithLeft(func() string { return "left" }),
+			WithProgress(func() (float64, string) { return 1.0, "right" }),
+		},
+			"left [======================================] right"},
+		{[]ProgressBarOption{
+			WithLeft(func() string { return "left" }),
+			WithProgress(func() (float64, string) { return -1, "right" }),
+		},
+			"left [" + strings.Repeat("-", 76) + "] right"},
+		{[]ProgressBarOption{
+			WithLeft(func() string { return "left" }),
+			WithProgress(func() (float64, string) { return 2, "right" }),
+		},
+			"left [" + strings.Repeat("=", 76) + "] right"},
```

Comment:
Should these two cases be fixed? Ideally, progress values should be clipped between `0=>x=>1` to avoid weird UI issues like this (currently the progress bar grows proportionally to this value).

Reply:
Hmm I'm a bit torn... On the one hand, I think we should clip the progress values between 0 and 1, because any other values don't make sense. On the other hand, if we pass return sub-0 or over-1 values from the rest of the k6 codebase, that's probably a bug, and it would be much easier to notice if we don't clip the progressbar values :sweat_smile: 

So... ¯\_(ツ)_/¯ maybe we should clamp the value between 0 and 1, but actually emit a warning if we get invalid values? @cuonglm @MStoykov , thoughts?

Reply:
Yeah, clamping+warning SGTM.

Reply:
Yeah I am for a warning and clamping as well ...

Reply:
👍

Reply:
See f9d9d611. Far from ideal...

It's missing the logging test because of the circular import caused by importing testutils.SimpleLogrusHook, which is also why Clampf was copied... :man_facepalming: The main problem is `lib/executors.go`. I tried moving that file to `lib/executor/` but it's a huge PITA to refactor, even with `gofmt -r`, and quickly gave up.

Also passing the logger everywhere is annoying, but it seemed more efficient / centralized than every `ProgressBar` managing its own instance. And I'm not sure what the log event will look like in cloud execution, but passed it for consistency. Maybe it shouldn't be?

Let me know.

Reply:
> See [f9d9d61](https://github.com/loadimpact/k6/commit/f9d9d611d756c41c5565967f62b402dd96bf4478). Far from ideal...

True, though copying the small `Clampf` doesn't really bother me very much... It's sort of an unfortunate reality with Go, lacking generics, you can't have a nice small library of such utility functions, so you end up copying the same code with different types. 

In this case our module structure is to blame though, and yeah, eventually we'd have to refactor things... maybe split the testutils package into 2, one part that's totally k6-agnostic and can be used anywhere, and another that would depend on `k6/lib` and be useful for testing some k6-specific things... Anyway, this is fine for now, I think.


> Also passing the logger everywhere is annoying, but it seemed more efficient / centralized than every `ProgressBar` managing its own instance. And I'm not sure what the log event will look like in cloud execution, but passed it for consistency. Maybe it shouldn't be?

Maybe just add a `WithLogger()` configurator to the `pb` package? Then you can just pass the logger that each executor has to it, when you create the progressbar? https://github.com/loadimpact/k6/blob/f6be35cfa006c56954d6821e0f94f4863f063e45/lib/executor/base_executor.go#L47-L49